File: core/src/main/java/javabot/dao/ChangeDao.java
package javabot.dao;

import com.antwerkz.sofia.Sofia;
import com.mongodb.WriteResult;
import javabot.dao.util.QueryParam;
import javabot.model.Change;
import javabot.model.Karma;
import javabot.model.criteria.ChangeCriteria;
import org.mongodb.morphia.query.Query;

import java.time.LocalDateTime;
import java.util.List;

public class ChangeDao extends BaseDao<Change> {
    public ChangeDao() {
        super(Change.class);
    }

    public void logChange(final String message) {
        final Change change = new Change();
        change.setMessage(message);
        change.setChangeDate(LocalDateTime.now());
        save(change);
    }

    public void logAdd(final String sender, final String key, final String value) {
        logChange(Sofia.factoidAdded(sender, key, value));
    }

    public boolean findLog(final String message) {
        ChangeCriteria criteria = new ChangeCriteria(ds);
        criteria.message().equal(message);
        return criteria.query().countAll() != 0;
    }

    public Long count(final Change filter) {
        return buildFindQuery(null, filter, true).countAll();
    }

    @SuppressWarnings({"unchecked"})
    public List<Change> getChanges(final QueryParam qp, final Change filter) {
        return buildFindQuery(qp, filter, true).asList();
    }

    private Query<Change> buildFindQuery(final QueryParam qp, final Change filter, final boolean count) {
        ChangeCriteria criteria = new ChangeCriteria(ds);
        if (filter.getId() != null) {
            criteria.id().equal(filter.getId());
        }
        if (filter.getMessage() != null) {
            criteria.message().contains(filter.getMessage());
        }
        if (filter.getChangeDate() != null) {
            criteria.changeDate(filter.getChangeDate());
        }
        criteria.changeDate().order(false);
        if (!count && qp != null) {
            criteria.query().offset(qp.getFirst());
            criteria.query().limit(qp.getCount());
        }
        return criteria.query();
    }

    public WriteResult deleteAll() {
        return ds.delete(ds.createQuery(Change.class));
    }
}

File: src/main/kotlin/javabot/dao/ChangeDao.kt
package javabot.dao

import com.antwerkz.sofia.Sofia
import com.google.inject.Inject
import com.mongodb.client.result.DeleteResult
import dev.morphia.Datastore
import dev.morphia.DeleteOptions
import dev.morphia.query.FindOptions
import dev.morphia.query.Query
import dev.morphia.query.Sort.descending
import dev.morphia.query.filters.Filters.eq
import dev.morphia.query.filters.Filters.regex
import java.time.LocalDateTime
import javabot.dao.util.QueryParam
import javabot.model.Change

class ChangeDao @Inject constructor(ds: Datastore) : BaseDao<Change>(ds, Change::class.java) {

    fun logFactoidAdded(sender: String, key: String, value: String, location: String) {
        save(Change(Sofia.factoidAdded(sender, key, value, location)))
    }

    fun logFactoidChanged(
        sender: String,
        key: String,
        oldValue: String,
        newValue: String,
        location: String,
    ) {
        save(Change(Sofia.factoidChanged(sender, key, oldValue, newValue, location)))
    }

    fun logFactoidRemoved(sender: String, key: String, value: String, location: String) {
        save(Change(Sofia.factoidRemoved(sender, key, value, location)))
    }

    fun logKarmaChanged(sender: String, target: String, value: Int, location: String) {
        save(Change(Sofia.karmaChanged(sender, target, value, location)))
    }

    fun logChangingLockedFactoid(nick: String, key: String, channel: String) {
        save(Change(Sofia.factoidChangingLocked(nick, key, channel)))
    }

    fun findLog(message: String): Boolean {
        val query = ds.find(Change::class.java).filter(eq("message", message))

        return query.count() != 0L
    }

    fun count(message: String?, date: LocalDateTime?): Long {
        return buildFindQuery(message, date).count()
    }

    fun getChanges(qp: QueryParam, message: String?, date: LocalDateTime?): List<Change> {
        return buildFindQuery(message, date)
            .iterator(FindOptions().skip(qp.first).limit(qp.count).sort(descending("changeDate")))
            .toList()
    }

    private fun buildFindQuery(message: String?, date: LocalDateTime?): Query<Change> {
        val query = ds.find(Change::class.java)
        if (message != null) {
            query.filter(regex("message").pattern(message))
        }
        if (date != null) {
            query.filter(eq("changeDate", date))
        }
        return query
    }

    fun deleteAll(): DeleteResult {
        return ds.find(Change::class.java).delete(DeleteOptions().multi(true))
    }
}
